### ISSUE TITLE
Mockk mockkStatic failure when using Robolectric 4.4

### ISSUE THREAD

**Issue by lcdsmao** [robolectric/robolectric#5994]
### Description
This issue is a copy of https://github.com/mockk/mockk/issues/480.

Mock static function failure when using Robolectric 4.4

### Steps to Reproduce

```kotlin
// Test target (pkg.Fookt):

class Foo
fun Foo.test() = "foo"

// Test code
@RunWith(AndroidJUnit4::class)
class FooTest {

  @Test
  fun test() {
    mockkStatic("pkg.FooKt")
    val foo: Foo = mockk()
    every { foo.test() } returns "bar"
    println(foo.test()) // Fail here
    // Remove @RunWith(AndroidJUnit4::class) will print bar
  }
}
```

### Robolectric & Android Version

* MockK version: 1.10.2
* OS: MacOS 10.15.6 (19G2021)
* Kotlin version: 1.4.10
* JDK version: 1.8
* Robolectric version: 4.4
* SDK: 28

**Comment by hoisie** [comment]
Just to be forthcoming, fixing this issue will likely be a code change in the MockK project, not in Robolectric. I'd be happy to provide help on the Robolectric side, but if you want a timely fix, I strongly encourage you (or someone else) to investigate the root cause yourself.

**Comment by lcdsmao** [comment]
Hi @hoisie.
I found using `mockkStatic(Class.forName("...").kotlin)` can mock static properly:

```kotlin
@RunWith(RobolectricTestRunner::class)
class MockkStaticTestCase {

    private val foo = mockk<Foo>()

    @After
    fun tearDown() {
        unmockkAll()
    }

    @Test
    fun `MockkStaticByClassName`() {
        // Failure
        mockkStatic("org.robolectric.integrationtests.mockk.StaticExtensionKt")
        every { foo.returnNumber() } returns 1
        assert(foo.returnNumber() == 1)
    }

    @Test
    fun `MockkStaticByClass`() {
        // Success
        mockkStatic(Class.forName("org.robolectric.integrationtests.mockk.StaticExtensionKt").kotlin)
        every { foo.returnNumber() } returns 1
        assert(foo.returnNumber() == 1)
    }
}
```

It seems the exclude for `io.mockk` in this PR #5754 affected the reflection call in mockk [InternalPlatformDsl.classForName](https://github.com/mockk/mockk/blob/54e29c1d98b19877d50a5a2d14426fa0ee4bfbe2/dsl/jvm/src/main/kotlin/io/mockk/InternalPlatformDsl.kt#L90).
Thanks.

**Comment by hoisie** [comment]
It is possible that #5754 may be exclucing too much of MockK from the Robolectric classloader. Instead of `.doNotAcquirePackage("io.mockk.")`, perhaps it should be replaced with something more specific to the dispatch-related classes of MockK. This requires some knowledge of the MockK internals, which unfortunately I do not have.

**Comment by Raibaz** [comment]
Hi there, MockK maintainer here.

It indeed looks like excluding the whole io.mockk. package is indeed preventing mockk from functioning correctly.

This being said, I'm in the same situation as @hoisie as I don't have enough knowledge about the Robolectric internals to help; however, I'm definitely open to explore any possibilities on how to fix this.

Let me know if there's any way I can help.

**Comment by hoisie** [comment]
Thanks for the comment @Raibaz. One of the main questions is which classes in MockK need to exist in the bootstrap class loader. This came up with Mockito as well:

https://github.com/mockito/mockito/issues/2015

For Mockito we excluded one class (the main dispatcher) from the Robolectric class loader but included the rest as part of the Robolectric class loader.

**Comment by Raibaz** [comment]
Judging from what was done with Mockito, it may be the case that excluding just `io.mockk.proxy.android.AndroidMockKDispatcher`, rather than the whole io.mockk package, could be enough.

Does this make sense to you?

**Comment by pocmo** [comment]
> `io.mockk.proxy.android.AndroidMockKDispatcher`, rather than the whole io.mockk package, could be enough.

That's something I tried locally (I came here from one of the linked mockk + robolectric 4 issues). Unfortunately, with just this exclude in a local Robolectric build, I end up with:

> java.lang.NoClassDefFoundError: Could not initialize class io.mockk.impl.JvmMockKGateway

**Comment by pocmo** [comment]
I can confirm that the patch from @ogolberg fixes most of the `ClassCastException` issues we have been seeing with Firefox for Android (after upgrading to Java 11 and bumping related dependencies). Not all of them are fixed, but it's down to a manageable size. :)

Is there any chance this patch could make it to a 4.4.x dot release? :)

**Comment by hoisie** [comment]
It is available now in 4.6-alpha-2. I can probably do a 4.5.x but 4.4.x is
pretty outdated at this point.

**Comment by pocmo** [comment]
I tested against the 4.5.x branch with the patch applied and I see the same improvements. A 4.5.x would work for us too and would definitely help us unblock the upgrade to Java 11 and newer Gradle tooling. :)

**Comment by utzcoz** [comment]
Looks like 4.6.x and later version fix this issue. Anyone has this problem can upgrade Robolectric to latest version to fix this problem. Feel free to reopen this issue if you think related patch must be ported to 4.4.x. If you find similar problem on latest Robolectric version, please file a new issue for it.
